File: crates/client/src/agent.rs
//! Driver for the server-side agent loop (`POST /v1/agent/runs`). Unlike the
//! single-request [`ChatBuilder::run_tools`] loop — which drives every
//! model->tool->model round-trip from the client — the gateway's agent-run
//! endpoint owns the loop server-side (mid-loop down-routing, judge-gated
//! summarize, substep cache). The client's job is therefore narrower: kick off
//! a run, and whenever the run pauses on a CLIENT (non-gateway) tool, execute
//! it via the caller's [`ToolExecutor`] and resume by POSTing the tool outputs,
//! until the run reaches a terminal answer.
//!
//! Cost differs from chat: the agent endpoint returns a single JSON [`Run`]
//! whose `usage` aggregates served cost across every server-side turn — there
//! are NO per-turn `x-tokentrimmer-*` headers — so the aggregate cost is read
//! from the response body, not the headers.

use serde::{Deserialize, Serialize};

use crate::{Client, Error, Message, MessageContent, Result, Tool, ToolCall, ToolExecutor};

/// Terminal (or paused) status of a run, mirroring the gateway's `RunStatus`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RunStatus {
    /// The model returned a final, tool-call-free answer.
    Completed,
    /// The loop stopped without a final answer (`max_turns` reached, or a
    /// client tool surfaced but the gateway has no run store to pause into).
    Incomplete,
    /// A completion turn errored server-side.
    Failed,
    /// The run paused on a client (non-gateway) tool and is awaiting the
    /// caller's tool outputs (resume via `POST .../tool_outputs`).
    RequiresAction,
}

impl RunStatus {
    /// True once the run can no longer be resumed (anything but `RequiresAction`).
    #[must_use]
    pub fn is_terminal(self) -> bool {
        !matches!(self, RunStatus::RequiresAction)
    }
}

/// Accumulated token usage + served cost across every turn of a run, mirroring
/// the gateway's `RunUsage`. `cost_usd` is the SUM of each turn's served cost
/// (`x-tokentrimmer-cost-usd`); it is the agent-loop analogue of
/// [`AggregateCost::cost_usd`](crate::AggregateCost::cost_usd).
#[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize)]
pub struct RunUsage {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    #[serde(default)]
    pub cost_usd: f64,
}

/// A run record returned by `POST /v1/agent/runs` and the resume endpoint — the
/// gateway's `Run` view. The full `messages` transcript is included so the
/// caller sees the whole model/tool exchange.
#[derive(Debug, Clone, Deserialize)]
pub struct Run {
    pub id: String,
    pub status: RunStatus,
    pub messages: Vec<Message>,
    pub turns: u32,
    pub usage: RunUsage,
    #[serde(default)]
    pub note: Option<String>,
    /// Summarizer measurement tax (USD) — a metering side-cost, never folded
    /// into `usage.cost_usd`. `None` when unmetered / no summarization.
    #[serde(default)]
    pub summarizer_tax_usd: Option<f64>,
}

impl Run {
    /// The final assistant text (`messages`' last assistant text block), if any.
    #[must_use]
    pub fn text(&self) -> Option<&str> {
        self.messages.iter().rev().find_map(|m| match m {
            Message::Assistant {
                content: Some(MessageContent::Text(t)),
                ..
            } => Some(t.as_str()),
            _ => None,
        })
    }

    /// The CLIENT tool calls the run is paused on: assistant `tool_calls` in the
    /// transcript that have NO answering `Message::Tool` block. The gateway
    /// answers its own (read-only gateway) tool calls inline before pausing, so
    /// the unanswered remainder are exactly the client tools the caller must run.
    ///
    /// Empty unless [`status`](Self::status) is `RequiresAction`.
    #[must_use]
    pub fn pending_tool_calls(&self) -> Vec<ToolCall> {
        let answered: std::collections::HashSet<&str> = self
            .messages
            .iter()
            .filter_map(|m| match m {
                Message::Tool { tool_call_id, .. } => Some(tool_call_id.as_str()),
                _ => None,
            })
            .collect();
        self.messages
            .iter()
            .filter_map(|m| match m {
                Message::Assistant { tool_calls, .. } => Some(tool_calls),
                _ => None,
            })
            .flatten()
            .filter(|tc| !answered.contains(tc.id.as_str()))
            .cloned()
            .collect()
    }
}

/// The result of driving an agent run to a terminal state (or hitting the
/// client-side resume cap).
#[derive(Debug, Clone)]
pub struct AgentOutcome {
    /// The final run record (its `status` is terminal unless the resume cap was
    /// hit while the run was still `RequiresAction`).
    pub run: Run,
    /// Number of `POST .../tool_outputs` resume round-trips the client made.
    pub resume_rounds: usize,
}

impl AgentOutcome {
    /// The run's final answer text, if any. See [`Run::text`].
    #[must_use]
    pub fn text(&self) -> Option<&str> {
        self.run.text()
    }
}

/// One tool output submitted to resume a paused run. Wire shape:
/// `{ "tool_call_id": "...", "output": "..." }`.
#[derive(Debug, Clone, Serialize)]
struct ToolOutput {
    tool_call_id: String,
    output: String,
}

/// Fluent builder for an agent run. See [`Client::agent`].
pub struct AgentBuilder<'a> {
    client: &'a Client,
    model: String,
    messages: Vec<Message>,
    tools: Vec<Tool>,
    /// Server-side per-run turn cap; `None` ⇒ the gateway default. Forwarded as
    /// the request's `max_turns` (the gateway clamps it to `[1, 32]`).
    max_turns: Option<u32>,
    tag: Option<String>,
    cost_limit: Option<f64>,
    interactive: bool,
    /// Client-side cap on resume (`tool_outputs`) round-trips before the driver
    /// returns the still-paused run, so a model that loops forever on client
    /// tools can't drive an unbounded number of resume requests.
    max_resume_rounds: usize,
}

impl<'a> AgentBuilder<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Self {
            client,
            model: String::new(),
            messages: Vec::new(),
            tools: Vec::new(),
            max_turns: None,
            tag: None,
            cost_limit: None,
            interactive: false,
            max_resume_rounds: 8,
        }
    }

    #[must_use]
    pub fn model(mut self, model: impl Into<String>) -> Self {
        self.model = model.into();
        self
    }
    #[must_use]
    pub fn messages(mut self, messages: Vec<Message>) -> Self {
        self.messages = messages;
        self
    }
    #[must_use]
    pub fn message(mut self, message: Message) -> Self {
        self.messages.push(message);
        self
    }
    /// Advertise function `tools` to the model. A run only pauses on tools the
    /// gateway does NOT execute itself (its read-only gateway tools run
    /// server-side); the ones you list here that aren't gateway tools are the
    /// ones your [`ToolExecutor`] will be asked to run.
    #[must_use]
    pub fn tools(mut self, tools: impl IntoIterator<Item = Tool>) -> Self {
        self.tools = tools.into_iter().collect();
        self
    }
    /// Server-side per-run turn cap (the gateway clamps to `[1, 32]`).
    #[must_use]
    pub fn max_turns(mut self, n: u32) -> Self {
        self.max_turns = Some(n);
        self
    }
    /// `X-TokenTrimmer-Tag` — free-form cost-attribution tag, forwarded on the
    /// initial create AND every resume request.
    #[must_use]
    pub fn tag(mut self, tag: impl Into<String>) -> Self {
        self.tag = Some(tag.into());
        self
    }
    /// `X-TokenTrimmer-Cost-Limit-Usd` — the gateway rejects with `402` if a
    /// turn's estimated cost exceeds `usd`.
    #[must_use]
    pub fn cost_limit(mut self, usd: f64) -> Self {
        self.cost_limit = Some(usd);
        self
    }
    /// Mark the run as **interactive** (a human is waiting): sends
    /// `X-TokenTrimmer-Interactive: 1`, clearing the gateway's advisory
    /// batch-eligibility marker.
    #[must_use]
    pub fn interactive(mut self) -> Self {
        self.interactive = true;
        self
    }
    /// Max client-side resume (`tool_outputs`) round-trips before the driver
    /// returns a still-paused run (default 8). A guard against a model that
    /// loops forever on client tools — independent of the server's `max_turns`.
    #[must_use]
    pub fn max_resume_rounds(mut self, n: usize) -> Self {
        self.max_resume_rounds = n;
        self
    }

    /// Drive the agent run to completion: create the run, then while it is
    /// paused on client tools, execute them via `executor` and resume — until a
    /// terminal status or the resume cap.
    ///
    /// # Errors
    /// [`Error::MissingModel`] if no model was set; [`Error::Request`] /
    /// [`Error::Status`] / [`Error::Decode`] on a gateway/transport failure of
    /// either the create or a resume call (a gateway failure aborts the run).
    /// Per-tool executor errors do NOT propagate — they are submitted back to
    /// the run as the tool's output so the model can recover.
    pub async fn run(self, executor: &(impl ToolExecutor + Sync)) -> Result<AgentOutcome> {
        if self.model.trim().is_empty() {
            return Err(Error::MissingModel);
        }
        let mut run = self.create().await?;
        let mut resume_rounds = 0usize;

        while run.status == RunStatus::RequiresAction && resume_rounds < self.max_resume_rounds {
            let pending = run.pending_tool_calls();
            // A `requires_action` run always has >=1 pending client tool; an
            // empty list would mean a server contract break. Stop rather than
            // POST an empty resume the gateway would reject (it requires the
            // outputs to exactly cover the pending ids).
            if pending.is_empty() {
                break;
            }
            let mut outputs = Vec::with_capacity(pending.len());
            for tc in &pending {
                let output = match executor
                    .call(&tc.function.name, &tc.function.arguments)
                    .await
                {
                    Ok(s) => s,
                    // Feed the error BACK as the tool output (mirrors the chat
                    // run_tools loop) so the model can react; never abort.
                    Err(e) => serde_json::json!({ "error": e.to_string() }).to_string(),
                };
                outputs.push(ToolOutput {
                    tool_call_id: tc.id.clone(),
                    output,
                });
            }
            run = self.resume(&run.id, &outputs).await?;
            resume_rounds += 1;
        }

        Ok(AgentOutcome { run, resume_rounds })
    }

    /// POST the initial `/v1/agent/runs` request.
    async fn create(&self) -> Result<Run> {
        let mut body = serde_json::json!({
            "model": self.model,
            "messages": self.messages,
            "stream": false,
        });
        if !self.tools.is_empty() {
            body["tools"] = serde_json::to_value(&self.tools).unwrap_or(serde_json::Value::Null);
        }
        if let Some(mt) = self.max_turns {
            body["max_turns"] = serde_json::json!(mt);
        }
        let url = format!("{}/v1/agent/runs", self.client.base);
        self.send_run(url, body).await
    }

    /// POST `/v1/agent/runs/{id}/tool_outputs` to resume a paused run.
    async fn resume(&self, run_id: &str, outputs: &[ToolOutput]) -> Result<Run> {
        let body = serde_json::json!({ "tool_outputs": outputs });
        let url = format!("{}/v1/agent/runs/{run_id}/tool_outputs", self.client.base);
        self.send_run(url, body).await
    }

    /// Shared POST helper: attach auth + tt headers, send, surface non-2xx as
    /// [`Error::Status`], decode the [`Run`] body.
    async fn send_run(&self, url: String, body: serde_json::Value) -> Result<Run> {
        let req = self
            .client
            .http
            .post(url)
            .bearer_auth(&self.client.key)
            .json(&body);
        let req =
            crate::apply_tt_headers(req, self.tag.as_deref(), self.cost_limit, self.interactive)?;
        let resp = req.send().await.map_err(Error::Request)?;
        let cost = crate::parse_cost(resp.headers());
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            return Err(Error::Status {
                status: status.as_u16(),
                body,
                cost: Box::new(cost),
            });
        }
        resp.json::<Run>().await.map_err(Error::Decode)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{async_trait, system, tool, user, Client};
    use httpmock::prelude::*;
    use serde_json::json;

    /// A canned client-tool executor returning a fixed string.
    struct Canned(&'static str);
    #[async_trait]
    impl ToolExecutor for Canned {
        async fn call(
            &self,
            _name: &str,
            _arguments: &str,
        ) -> std::result::Result<String, Box<dyn std::error::Error + Send + Sync>> {
            Ok(self.0.to_string())
        }
    }

    struct AlwaysErr;
    #[async_trait]
    impl ToolExecutor for AlwaysErr {
        async fn call(
            &self,
            _name: &str,
            _arguments: &str,
        ) -> std::result::Result<String, Box<dyn std::error::Error + Send + Sync>> {
            Err("tool exploded".into())
        }
    }

    /// A `requires_action` Run paused on a client tool `do_thing` (id `call_1`),
    /// whose assistant tool_call is unanswered in the transcript.
    fn requires_action_run() -> serde_json::Value {
        json!({
            "id": "11111111-1111-1111-1111-111111111111",
            "status": "requires_action",
            "turns": 1,
            "usage": { "prompt_tokens": 10, "completion_tokens": 4, "cost_usd": 0.0002 },
            "messages": [
                { "role": "user", "content": "do it" },
                { "role": "assistant", "content": null, "tool_calls": [
                    { "id": "call_1", "type": "function",
                      "function": { "name": "do_thing", "arguments": "{\"x\":1}" } }
                ]}
            ]
        })
    }

    /// The terminal Run returned after resume: the tool output was appended and
    /// the model produced a final text answer. Aggregate cost spans both turns.
    fn completed_run_after_resume() -> serde_json::Value {
        json!({
            "id": "11111111-1111-1111-1111-111111111111",
            "status": "completed",
            "turns": 2,
            "usage": { "prompt_tokens": 22, "completion_tokens": 9, "cost_usd": 0.0005 },
            "messages": [
                { "role": "user", "content": "do it" },
                { "role": "assistant", "content": null, "tool_calls": [
                    { "id": "call_1", "type": "function",
                      "function": { "name": "do_thing", "arguments": "{\"x\":1}" } }
                ]},
                { "role": "tool", "tool_call_id": "call_1", "content": "did the thing" },
                { "role": "assistant", "content": "All done." }
            ]
        })
    }

    #[test]
    fn pending_tool_calls_are_the_unanswered_assistant_calls() {
        let run: Run = serde_json::from_value(requires_action_run()).unwrap();
        let pending = run.pending_tool_calls();
        assert_eq!(pending.len(), 1);
        assert_eq!(pending[0].id, "call_1");
        assert_eq!(pending[0].function.name, "do_thing");

        // After the answering Tool block is present, nothing is pending.
        let done: Run = serde_json::from_value(completed_run_after_resume()).unwrap();
        assert!(done.pending_tool_calls().is_empty());
        assert_eq!(done.text(), Some("All done."));
        assert!(done.status.is_terminal());
    }

    #[tokio::test]
    async fn run_executes_client_tool_then_resumes_to_completion() {
        let server = MockServer::start_async().await;
        // Resume endpoint (more specific path) → terminal completed run.
        let resume = server.mock(|when, then| {
            when.method(POST)
                .path("/v1/agent/runs/11111111-1111-1111-1111-111111111111/tool_outputs")
                .body_includes("did the thing")
                .body_includes("call_1");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(completed_run_after_resume());
        });
        // Create endpoint → paused requires_action run.
        let create = server.mock(|when, then| {
            when.method(POST).path("/v1/agent/runs");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(requires_action_run());
        });

        let client = Client::new(server.base_url(), "k");
        let out = client
            .agent()
            .model("gpt-4o-mini")
            .message(system("be helpful"))
            .message(user("do it"))
            .tools(vec![tool("do_thing", "does", json!({"type":"object"}))])
            .run(&Canned("did the thing"))
            .await
            .unwrap();

        create.assert();
        resume.assert();
        assert_eq!(out.resume_rounds, 1);
        assert_eq!(out.run.status, RunStatus::Completed);
        assert_eq!(out.text(), Some("All done."));
        // Aggregate cost is read from the terminal run body, not headers.
        assert!((out.run.usage.cost_usd - 0.0005).abs() < 1e-9);
        assert_eq!(out.run.usage.prompt_tokens, 22);
    }

    #[tokio::test]
    async fn run_returns_immediately_when_first_response_is_terminal() {
        // The gateway ran the whole loop server-side (only gateway tools) and
        // returned a completed run — no client tool, no resume.
        let server = MockServer::start_async().await;
        let m = server.mock(|when, then| {
            when.method(POST).path("/v1/agent/runs");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(json!({
                    "id": "abc", "status": "completed", "turns": 3,
                    "usage": { "prompt_tokens": 30, "completion_tokens": 12, "cost_usd": 0.0009 },
                    "messages": [
                        { "role": "user", "content": "hi" },
                        { "role": "assistant", "content": "Answer." }
                    ]
                }));
        });
        let client = Client::new(server.base_url(), "k");
        let out = client
            .agent()
            .model("gpt-4o-mini")
            .message(user("hi"))
            .run(&Canned("unused"))
            .await
            .unwrap();
        m.assert();
        assert_eq!(out.resume_rounds, 0);
        assert_eq!(out.text(), Some("Answer."));
    }

    #[tokio::test]
    async fn run_stops_at_resume_cap_when_run_keeps_pausing() {
        // The gateway never finishes: every create/resume returns the SAME
        // requires_action run. With max_resume_rounds(2) the driver makes the
        // create + exactly 2 resumes, then returns the still-paused run.
        let server = MockServer::start_async().await;
        // Both the create AND every resume return the same paused run; one mock
        // per path (httpmock `path` is an exact match).
        server.mock(|when, then| {
            when.method(POST)
                .path("/v1/agent/runs/11111111-1111-1111-1111-111111111111/tool_outputs");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(requires_action_run());
        });
        server.mock(|when, then| {
            when.method(POST).path("/v1/agent/runs");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(requires_action_run());
        });
        let client = Client::new(server.base_url(), "k");
        let out = client
            .agent()
            .model("gpt-4o-mini")
            .message(user("loop"))
            .tools(vec![tool("do_thing", "does", json!({"type":"object"}))])
            .max_resume_rounds(2)
            .run(&Canned("again"))
            .await
            .unwrap();
        assert_eq!(out.resume_rounds, 2);
        assert_eq!(out.run.status, RunStatus::RequiresAction);
    }

    #[tokio::test]
    async fn run_feeds_executor_error_back_as_tool_output() {
        let server = MockServer::start_async().await;
        // Resume carrying the ERROR output → terminal recovered run.
        let resume = server.mock(|when, then| {
            when.method(POST)
                .path("/v1/agent/runs/11111111-1111-1111-1111-111111111111/tool_outputs")
                .body_includes("exploded");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(completed_run_after_resume());
        });
        server.mock(|when, then| {
            when.method(POST).path("/v1/agent/runs");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(requires_action_run());
        });
        let client = Client::new(server.base_url(), "k");
        let out = client
            .agent()
            .model("gpt-4o-mini")
            .message(user("go"))
            .tools(vec![tool("do_thing", "does", json!({"type":"object"}))])
            .run(&AlwaysErr)
            .await
            .unwrap();
        resume.assert(); // the error was submitted, the loop did NOT abort
        assert_eq!(out.run.status, RunStatus::Completed);
    }

    #[tokio::test]
    async fn run_without_model_errors_before_any_request() {
        let client = Client::new("http://127.0.0.1:1", "k");
        let err = client
            .agent()
            .message(user("hi"))
            .run(&Canned("x"))
            .await
            .unwrap_err();
        assert!(matches!(err, Error::MissingModel), "{err:?}");
    }

    #[tokio::test]
    async fn run_propagates_gateway_error_on_create() {
        let server = MockServer::start_async().await;
        server.mock(|when, then| {
            when.method(POST).path("/v1/agent/runs");
            then.status(500).body("boom");
        });
        let client = Client::new(server.base_url(), "k");
        let result = client
            .agent()
            .model("gpt-4o-mini")
            .message(user("go"))
            .run(&Canned("x"))
            .await;
        assert!(matches!(result, Err(Error::Status { status: 500, .. })));
    }

    #[tokio::test]
    async fn run_propagates_gateway_error_on_resume() {
        // Create pauses fine; the resume fails → the error surfaces (run aborts).
        let server = MockServer::start_async().await;
        server.mock(|when, then| {
            when.method(POST)
                .path("/v1/agent/runs/11111111-1111-1111-1111-111111111111/tool_outputs");
            then.status(503).body("store unavailable");
        });
        server.mock(|when, then| {
            when.method(POST).path("/v1/agent/runs");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(requires_action_run());
        });
        let client = Client::new(server.base_url(), "k");
        let result = client
            .agent()
            .model("gpt-4o-mini")
            .message(user("go"))
            .tools(vec![tool("do_thing", "does", json!({"type":"object"}))])
            .run(&Canned("x"))
            .await;
        assert!(matches!(result, Err(Error::Status { status: 503, .. })));
    }

    #[tokio::test]
    async fn run_forwards_tag_and_max_turns() {
        let server = MockServer::start_async().await;
        let m = server.mock(|when, then| {
            when.method(POST)
                .path("/v1/agent/runs")
                .header("x-tokentrimmer-tag", "proj-x")
                .body_includes("\"max_turns\":4");
            then.status(200)
                .header("content-type", "application/json")
                .json_body(json!({
                    "id": "z", "status": "completed", "turns": 1,
                    "usage": { "prompt_tokens": 1, "completion_tokens": 1, "cost_usd": 0.0 },
                    "messages": [{ "role": "assistant", "content": "ok" }]
                }));
        });
        let client = Client::new(server.base_url(), "k");
        client
            .agent()
            .model("gpt-4o-mini")
            .message(user("hi"))
            .tag("proj-x")
            .max_turns(4)
            .run(&Canned("x"))
            .await
            .unwrap();
        m.assert();
    }
}

File: sdk-python/tokentrimmer/agent.py
"""Driver for the server-side agent loop (``POST /v1/agent/runs``).

Unlike a single chat completion — where the caller drives every
model->tool->model round-trip from the client — the Gateway's agent-run
endpoint owns the loop server-side (mid-loop down-routing, judge-gated
summarize, substep cache). The client's job is therefore narrower: kick off a
run, and whenever the run pauses on a CLIENT (non-gateway) tool, execute it via
the caller's ``executor`` and resume by POSTing the tool outputs, until the run
reaches a terminal answer.

This mirrors the Rust ``tt-client`` driver (crates/client/src/agent.rs).

Cost differs from chat: the agent endpoint returns a single JSON :class:`Run`
whose ``usage`` aggregates served cost across every server-side turn — there are
NO per-turn ``x-tokentrimmer-*`` headers — so the aggregate cost is read from the
response body (``run.usage.cost_usd``), not the headers.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional
from urllib.parse import quote

import httpx

# A caller-supplied client-tool executor: ``(name, arguments) -> output``.
# ``arguments`` is the raw JSON-string the model produced; ``output`` is the raw
# string fed back to the model as the tool result. Raising is allowed — the
# driver catches it and feeds the error back as the tool output (never aborts).
Executor = Callable[[str, str], str]

# Resume (``tool_outputs``) round-trip cap, matching the Rust driver default. A
# guard against a model that loops forever on client tools — independent of the
# server-side ``max_turns``.
DEFAULT_MAX_RESUME_ROUNDS = 8


@dataclass(frozen=True)
class RunUsage:
    """Accumulated token usage + served cost across every turn of a run.

    ``cost_usd`` is the SUM of each turn's served cost; it is the agent-loop
    analogue of a chat response's ``.tt.cost_usd``.
    """

    prompt_tokens: int = 0
    completion_tokens: int = 0
    cost_usd: float = 0.0

    @classmethod
    def _from_payload(cls, data: Optional[Dict[str, Any]]) -> "RunUsage":
        data = data or {}
        return cls(
            prompt_tokens=int(data.get("prompt_tokens", 0)),
            completion_tokens=int(data.get("completion_tokens", 0)),
            cost_usd=float(data.get("cost_usd", 0.0)),
        )


@dataclass(frozen=True)
class Run:
    """A run record returned by ``POST /v1/agent/runs`` and the resume endpoint.

    The full ``messages`` transcript is included so the caller sees the whole
    model/tool exchange. ``status`` is one of ``completed`` / ``incomplete`` /
    ``failed`` / ``requires_action``; only ``requires_action`` is resumable.
    """

    id: str
    status: str
    messages: List[Dict[str, Any]]
    turns: int
    usage: RunUsage
    note: Optional[str] = None
    summarizer_tax_usd: Optional[float] = None

    @classmethod
    def _from_payload(cls, data: Dict[str, Any]) -> "Run":
        return cls(
            id=str(data.get("id", "")),
            status=str(data.get("status", "")),
            messages=list(data.get("messages", [])),
            turns=int(data.get("turns", 0)),
            usage=RunUsage._from_payload(data.get("usage")),
            note=data.get("note"),
            summarizer_tax_usd=data.get("summarizer_tax_usd"),
        )

    @property
    def is_terminal(self) -> bool:
        """True once the run can no longer be resumed (anything but ``requires_action``)."""
        return self.status != "requires_action"

    @property
    def text(self) -> Optional[str]:
        """The final assistant text (last assistant message's string content), if any."""
        for msg in reversed(self.messages):
            if msg.get("role") == "assistant":
                content = msg.get("content")
                if isinstance(content, str):
                    return content
        return None

    def pending_tool_calls(self) -> List[Dict[str, Any]]:
        """The CLIENT tool calls the run is paused on.

        These are assistant ``tool_calls`` in the transcript that have NO
        answering ``tool`` message. The Gateway answers its own (read-only)
        gateway tool calls inline before pausing, so the unanswered remainder
        are exactly the client tools the caller must run. Empty unless
        ``status == "requires_action"``.
        """
        answered = {
            msg.get("tool_call_id")
            for msg in self.messages
            if msg.get("role") == "tool"
        }
        pending: List[Dict[str, Any]] = []
        for msg in self.messages:
            if msg.get("role") != "assistant":
                continue
            for call in msg.get("tool_calls") or []:
                if call.get("id") not in answered:
                    pending.append(call)
        return pending


@dataclass(frozen=True)
class AgentOutcome:
    """The result of driving an agent run to a terminal state (or the resume cap)."""

    run: Run
    #: Number of ``POST .../tool_outputs`` resume round-trips the driver made.
    resume_rounds: int

    @property
    def text(self) -> Optional[str]:
        """The run's final answer text, if any. See :attr:`Run.text`."""
        return self.run.text

    @property
    def usage(self) -> RunUsage:
        """Aggregate token usage + served cost across the whole run."""
        return self.run.usage

    @property
    def messages(self) -> List[Dict[str, Any]]:
        """The full final transcript of the run."""
        return self.run.messages


def _executor_error_output(exc: BaseException) -> str:
    """Render an executor exception as the JSON tool output fed back to the model.

    Mirrors the Rust driver's ``{"error": "..."}`` payload so the model can react
    to a failed tool rather than the whole run aborting.
    """
    return json.dumps({"error": str(exc)})


class Agent:
    """Server-side agent-loop driver, exposed as ``client.agent``.

    Reuses the SDK's existing base URL, API key, and underlying httpx client (no
    new HTTP dependency, and the same transport the chat path uses — so it is
    mockable with the same test harness).
    """

    def __init__(self, client: Any) -> None:
        # The owning ``TokenTrimmer`` (an ``openai.OpenAI`` subclass). We read
        # ``base_url`` / ``api_key`` and the underlying sync httpx client off it.
        self._client = client

    @property
    def _base(self) -> str:
        # ``client.base_url`` is an httpx.URL ending in a slash (e.g.
        # ``http://host/v1/``); strip it so we can join clean path segments.
        return str(self._client.base_url).rstrip("/")

    @property
    def _http(self) -> httpx.Client:
        # The OpenAI SDK's configured sync httpx client (respx-mockable).
        return self._client._client  # type: ignore[no-any-return]

    def run(
        self,
        *,
        model: str,
        messages: List[Dict[str, Any]],
        executor: Executor,
        tools: Optional[List[Dict[str, Any]]] = None,
        max_turns: Optional[int] = None,
        max_resume_rounds: int = DEFAULT_MAX_RESUME_ROUNDS,
        tt_tag: Optional[str] = None,
    ) -> AgentOutcome:
        """Drive the agent run to completion.

        Create the run, then while it is paused on client tools, execute them via
        ``executor`` and resume — until a terminal status or the resume cap.

        :param model: the model id to route the run with (required).
        :param messages: the OpenAI-shaped conversation to seed the run.
        :param executor: ``(name, arguments) -> output``; raising is fed back as
            the tool's output (the run does not abort). See :data:`Executor`.
        :param tools: function tools advertised to the model (OpenAI shape). A
            run only pauses on tools the Gateway does NOT execute itself.
        :param max_turns: server-side per-run turn cap (the Gateway clamps to
            ``[1, 32]``); ``None`` uses the Gateway default.
        :param max_resume_rounds: client-side cap on resume round-trips
            (default 8) before the driver returns a still-paused run.
        :param tt_tag: ``X-TokenTrimmer-Tag`` cost-attribution tag, forwarded on
            the create AND every resume request.
        :raises ValueError: if ``model`` is empty.
        :raises httpx.HTTPStatusError: on a gateway/transport failure of the
            create or a resume call (a gateway failure aborts the run).
        """
        if not model or not model.strip():
            raise ValueError("model must be a non-empty string")

        run = self._create(model, messages, tools, max_turns, tt_tag)
        resume_rounds = 0

        while run.status == "requires_action" and resume_rounds < max_resume_rounds:
            pending = run.pending_tool_calls()
            # A ``requires_action`` run always has >=1 pending client tool; an
            # empty list would mean a server contract break. Stop rather than
            # POST an empty resume the Gateway would reject.
            if not pending:
                break
            outputs = []
            for call in pending:
                fn = call.get("function") or {}
                name = str(fn.get("name", ""))
                arguments = str(fn.get("arguments", ""))
                try:
                    output = executor(name, arguments)
                except Exception as exc:  # noqa: BLE001 - feed error back, never abort
                    output = _executor_error_output(exc)
                outputs.append({"tool_call_id": call.get("id"), "output": output})
            run = self._resume(run.id, outputs, tt_tag)
            resume_rounds += 1

        return AgentOutcome(run=run, resume_rounds=resume_rounds)

    def export_transcript(self, run_id: str) -> Run:
        """Export a still-retained paused/resumed transcript.

        The Gateway returns 404 after the one-hour Redis record expires or is
        deleted. Durable run metadata remains available through run history.
        """
        url = self._transcript_url(run_id)
        resp = self._http.get(url, headers=self._headers())
        resp.raise_for_status()
        return Run._from_payload(resp.json())

    def delete_transcript(self, run_id: str) -> None:
        """Idempotently erase retained transcript/resume state only."""
        url = self._transcript_url(run_id)
        resp = self._http.delete(url, headers=self._headers())
        resp.raise_for_status()

    def _create(
        self,
        model: str,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Dict[str, Any]]],
        max_turns: Optional[int],
        tt_tag: Optional[str],
    ) -> Run:
        body: Dict[str, Any] = {"model": model, "messages": messages, "stream": False}
        if tools:
            body["tools"] = tools
        if max_turns is not None:
            body["max_turns"] = max_turns
        return self._send(f"{self._base}/agent/runs", body, tt_tag)

    def _resume(
        self, run_id: str, outputs: List[Dict[str, Any]], tt_tag: Optional[str]
    ) -> Run:
        body = {"tool_outputs": outputs}
        url = f"{self._base}/agent/runs/{run_id}/tool_outputs"
        return self._send(url, body, tt_tag)

    def _send(self, url: str, body: Dict[str, Any], tt_tag: Optional[str]) -> Run:
        headers = self._headers()
        if tt_tag is not None:
            headers["X-TokenTrimmer-Tag"] = str(tt_tag)
        resp = self._http.post(url, json=body, headers=headers)
        resp.raise_for_status()
        return Run._from_payload(resp.json())

    def _headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self._client.api_key}",
            "Content-Type": "application/json",
        }

    def _transcript_url(self, run_id: str) -> str:
        if not isinstance(run_id, str) or not run_id.strip():
            raise ValueError("run_id must be a non-empty string")
        return f"{self._base}/agent/runs/{quote(run_id, safe='')}/transcript"
